gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/UserQuotaManager.java

```diff
@@ -0,0 +1,198 @@
+        requesterCheck = requesterCheck && thisRequesterCheck;
+        if (!thisRequesterCheck) {
+          requesterMessage.append(String.format(
+              "Quota exceeded for requester %s on executor %s : quota=%s, requests above quota=%d%n",
+              requester, specExecutorUri, getQuotaForUser(requester), Math.abs(userQuotaIncrement)-getQuotaForUser(requester)));
+        }
+      }
+    }
+
+    // Throw errors for reach quota at the end to avoid inconsistent job counts
+    if ((!proxyUserCheck || !requesterCheck) && !onInit) {
+      // roll back the increased counts in this block
+      String userKey = DagManagerUtils.getUserQuotaKey(proxyUser, dagNode);
+      decrementQuotaUsage(proxyUserToJobCount, userKey);
+      decrementQuotaUsageForUsers(usersQuotaIncrement);
+      runningDagIds.remove(DagManagerUtils.generateDagId(dagNode));
+      throw new IOException(requesterMessage.toString());
+    }
+  }
+
+  /**
+   * Increment quota by one for the given map and key.
+   * @return a negative number if quota is already reached for this user
+   *         a positive number if the quota is not reached for this user
+   *         the absolute value of the number is the used quota before this increment request
+   *         0 if quota usage is not changed
+   */
+  private int incrementJobCountAndCheckUserQuota(Map<String, Integer> quotaMap, String user, Dag.DagNode<JobExecutionPlan> dagNode) {
+    String key = DagManagerUtils.getUserQuotaKey(user, dagNode);
+
+    // Only increment job count for first attempt, since job is considered running between retries
+    if (dagNode.getValue().getCurrentAttempts() != 1) {
+      return 0;
+    }
+
+    Integer currentCount;
+    // Modifications must be thread safe since DAGs on DagManagerThreads may update the quota for the same user
+    do {
+      currentCount = quotaMap.get(key);
+    } while (currentCount == null ? quotaMap.putIfAbsent(key, 1) != null : !quotaMap.replace(key, currentCount, currentCount + 1));
+
+    if (currentCount == null) {
+      currentCount = 0;
+    }
+
+    if (currentCount >= getQuotaForUser(user)) {
+      return -currentCount; // increment must have crossed the quota
+    } else {
+      return currentCount;
+    }
+  }
+
+  /**
+   * Decrement the quota by one for the proxy user and requesters corresponding to the provided {@link Dag.DagNode}.
+   * Returns true if the dag existed in the set of running dags and was removed successfully
+   */
+  public boolean releaseQuota(Dag.DagNode<JobExecutionPlan> dagNode) {
+    Boolean val = runningDagIds.remove(DagManagerUtils.generateDagId(dagNode));
+    if (val == null) {
+      return false;
+    }
+    String proxyUser = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(), AzkabanProjectConfig.USER_TO_PROXY, null);
+    if (proxyUser != null) {
+      String proxyUserKey = DagManagerUtils.getUserQuotaKey(proxyUser, dagNode);
+      decrementQuotaUsage(proxyUserToJobCount, proxyUserKey);
```

Comment:
QQ: do we need to surround this call with try as well, same as line 165?

Reply:
Yeah that's a good point, the original code didn't have a try/catch for proxyUser, I'm thinking because proxyuser always exist but service requester doesn't always exist? Otherwise the other functions here do not throw IOException so I'm not sure when it would be thrown at all

Reply:
Ah I see it now, it's thrown in `RequesterService.deserialize(serializedRequesters)` so it's not needed in this line